### PR TITLE
Add buildtask and dev URL permissions

### PR DESCRIPTION
**Pull request by andrewandante**
Introduces `canInit()` methods on Development admin controllers, and respects `canView()` permission checks on `BuildTask`s

## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/10852

### CODE REVIEW

**Comment by GuySartorelli** [comment]
@andrewandante Feel free to `@` me when this is ready - it looks like a great idea.

**Comment by andrewandante** [comment]
Thanks @GuySartorelli - I think this is in a good space, aside from the test. I'm having a bit of trouble turning off the `allow_all_cli` param, any advice?

While devving this I've noticed that it needs a companion PR `/dev/graphql` so I'll bundle that together as well

**Comment by GuySartorelli** [comment]
I think the problem is the environment type.
By adding this to `DevelopmentAdmin::canViewAll()` we can see that the `Director::isDev()` check is evaluating true, and the cli check is evaluating false.
```php
Debug::dump([
    'requested' => $requestedDevBuild,
    'isDev' => Director::isDev(),
    'CLI-and-allowed' => (Director::is_cli() && $allowAllCLI),
    'admin' => Permission::check("ADMIN"),
    'all-dev' => Permission::check("ALL_DEV_ADMIN"),
]);
```

Looks like just setting the environment variable isn't enough to flip the switch, so we need to call `setEnvironment()` on the kernel

```php
DevelopmentAdmin::config()->set('allow_all_cli', false);
$kernel = Injector::inst()->get(Kernel::class);
$env = $kernel->getEnvironment();
$kernel->setEnvironment(Kernel::LIVE);
try {
    // ..... do all your test stuff
} finally {
    $kernel->setEnvironment($env);
}
```

File: src/Dev/DevBuildController.php
Comment (michalkleiner) [inline src/Dev/DevBuildController.php]: (just thinking out loud, feel free to disregard)

Somehow the 'Can execute a Dev/Build' is pulling my ears, but I can't think of a better way.. dev/build command? Or perhaps 'Can execute the build command'? We all know it's on the dev controller, and therefore it became known as such.. if we look at e.g. the tasks command, listing all the tasks, we don't say dev/tasks when we talk about it..

```suggestion
                'name' => _t(__CLASS__ . '.CAN_DEV_BUILD_DESCRIPTION', 'Can execute dev/build'),
                'help' => _t(__CLASS__ . '.CAN_DEV_BUILD_HELP', 'Can run the build command (/dev/build).'),
```

File: src/Dev/DevelopmentAdmin.php
Comment (michalkleiner) [inline src/Dev/DevelopmentAdmin.php]: If the help text is required it might as well be the same as the description (vs just swapping two words around).

```suggestion
                'name' => _t(__CLASS__ . '.ALL_DEV_ADMIN_DESCRIPTION', 'Can view and run all /dev endpoints'),
                'help' => _t(__CLASS__ . '.ALL_DEV_ADMIN_HELP', 'Can view and run all endpoints at /dev).'),
```

**Comment by michalkleiner** [comment]
Good job, I like the idea 👍

File: src/Dev/DevBuildController.php
Comment (GuySartorelli) [reply src/Dev/DevBuildController.php]: > if we look at e.g. the tasks command, listing all the tasks, we don't say dev/tasks when we talk about it..

I do say dev/tasks when I talk about it :p But I also agree that "a Dev/Build" feels a little clunky, especially since "(/dev/build)" is there too. I like "the build command (/dev/build)"

Comment (andrewandante) [reply src/Dev/DevBuildController.php]: Yeah nice, I agree. Must admit I'd forgotten to revisit those strings after my initial pass through 😅

**Comment by andrewandante** [comment]
Thanks team, updated the wording based on your recommendations 👍

File: src/Dev/DevConfigController.php
Comment (GuySartorelli) [inline src/Dev/DevConfigController.php]:
```suggestion
                'name' => _t(__CLASS__ . '.CAN_DEV_CONFIG_DESCRIPTION', 'Can see /dev/config'),
                'help' => _t(__CLASS__ . '.CAN_DEV_CONFIG_HELP', 'Can see all application configuration (/dev/config).'),
```

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [inline src/Dev/DevelopmentAdmin.php]:
```suggestion
                'sort' => 50
```
Doesn't have to be 50, but this should be in the list _before_ all of the other dev/* permissions so it needs a lower sort than the others.
It might also make sense to give the rest of them specifically different sort values so they're in a determined order, but I won't hold up merging if you decide not to do that. I'm not really sure what order would be best myself so I'm not fussed.

File: src/Dev/DevBuildController.php
Comment (GuySartorelli) [inline src/Dev/DevBuildController.php]:
```suggestion
                'name' => _t(__CLASS__ . '.CAN_DEV_BUILD_DESCRIPTION', 'Can execute /dev/build'),
```
The global override says "Can view and run all /dev endpoints" so these should all label themselves as `/dev/*`

File: src/Dev/TaskRunner.php
Comment (GuySartorelli) [inline src/Dev/TaskRunner.php]:
```suggestion
                'name' => _t(__CLASS__ . '.BUILDTASK_CAN_RUN_DESCRIPTION', 'Can view and run /dev/tasks'),
                'help' => _t(__CLASS__ . '.BUILDTASK_CAN_RUN_HELP', 'Can view and run Build Tasks (/dev/tasks).'),
```
Consistent with the others

Comment (GuySartorelli) [reply src/Dev/TaskRunner.php]: I'm also now seeing that this permission overrides canView checks for the tasks - the language should be updated to reflect that. It's not clear right now that:
1. Some people may be able to view build tasks without this permission
2. Anyone with this permission will bypass any explicit canView checks on `BuildTask`s

File: src/ORM/DatabaseAdmin.php
Comment (GuySartorelli) [inline src/ORM/DatabaseAdmin.php]: Please put this in a `canInit()` method to be consistent with the other dev controllers - especially since `DevelopmentAdmin` is looking for that method.

File: src/Dev/TaskRunner.php
Comment (GuySartorelli) [inline src/Dev/TaskRunner.php]:
```suggestion
            || Permission::check(['ADMIN', 'ALL_DEV_ADMIN', 'BUILDTASK_CAN_RUN'])
```
This is a little tidier I think. Please make this change (single quotes, and a single check with an array) to all of these new permission checks.

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [inline src/Dev/DevelopmentAdmin.php:182]: If this is the replacement for `get_links`, please deprecate the old method - but in any case, why not just put this logic in the existing method?

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [inline src/Dev/DevelopmentAdmin.php:191]: Why is this necessary? It isn't in the `get_links` implementation.

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [inline src/Dev/DevelopmentAdmin.php:198]:
```suggestion
                    if ($controllerSingleton->hasMethod('canInit') && !$controllerSingleton->canInit()) {
                        continue;
                    }
```

I think if there's no `canInit()` method we should assume that access is allowed, rather than the opposite. If someone has implemented their own custom dev controller, they may have permission checks in `init()` which we're currently ignoring with this implementation.
I think.
Feel free to let me know if I've missed something there.

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [inline src/Dev/DevelopmentAdmin.php]:
```suggestion
                    $controllerSingleton = Injector::inst()->get($registeredController['controller']);
```
I was going to leave this alone but you've used this syntax in `TaskRunner` and I prefer it to the `singleton()` function.

File: src/Dev/TaskRunner.php
Comment (GuySartorelli) [inline src/Dev/TaskRunner.php]: Just running through the logic here for my own benefit more than anything.
- If the current user is admin or otherwise has permissions to view all tasks, canView on the task doesn't matter.
- If the current user can't view _all_ tasks (which before this PR basically means they're not admin), they can't view any task _unless_ that task explicitly implements a canView check.
- There's no (built-in) reason right now for for any `BuildTask` in the wild to have implemented a canView check, so theoretically this won't result in anyone being able to access tasks they couldn't access before - nor will it restrict access from tasks they could view in the past.

My one question I guess is whether this is the way around we want it - do we want a canView check to mean "if you don't have the /dev/tasks permission, maybe you can view/run this task", or do we want it to mean "even if you have the /dev/tasks permission, you might not be allowed to view/run this task"?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
From
screenshot of the permissions before changes are made
To
screenshot of the permissions after changes are made

Even then the language is a little bit inconsistent - we should probably swap "see" to "view", "run" to "execute", and "access" to "view and execute" (or change execute to run, either way so long as it's consistent)

File: src/Dev/TaskRunner.php
Comment (andrewandante) [reply src/Dev/TaskRunner.php]: Yep you've got that logic right - my thinking was to make sure we weren't changing any of the current behaviour quite yet, but providing the interface to.

The use-case I've seen for this is the former - "I don't want this user to be an admin, but I want them to be able to run GenerateTheTPSReportTask".

I'm not certain the other way round has a use-case I can think of - at least, you'd be able to craft a combination of permissions where you could achieve the same with this arrangement.

Comment (GuySartorelli) [reply src/Dev/TaskRunner.php]: Yup that makes sense

File: src/Dev/DevelopmentAdmin.php
Comment (andrewandante) [reply src/Dev/DevelopmentAdmin.php:198]: The problem with this interpretation is that now anyone can see those listed at `/dev` where before it was limited to `ADMIN` - I'm not super into either interpretation, to be honest, but I erred towards secure by default

File: src/Dev/DevelopmentAdmin.php
Comment (andrewandante) [reply src/Dev/DevelopmentAdmin.php:182]: `get_links()` is a static method and thus can't call `$this->canViewAll()` - which needs to access `$this->getRequest()` so can't be static itself

File: src/Dev/DevelopmentAdmin.php
Comment (andrewandante) [reply src/Dev/DevelopmentAdmin.php:191]: Otherwise we try to call `->hasMethod()` on `null`

Comment (GuySartorelli) [reply src/Dev/DevelopmentAdmin.php:191]: Let me step back a bit then - in what scenario are we getting a value in `$registeredController['controller']` which isn't a class?

Comment (andrewandante) [reply src/Dev/DevelopmentAdmin.php:191]: [In the tests](https://github.com/silverstripe/silverstripe-framework/blob/5/tests/php/Dev/DevAdminControllerTest.php#L32), actually 😅

File: src/ORM/DatabaseAdmin.php
Comment (andrewandante) [reply src/ORM/DatabaseAdmin.php]: Yeah this is interesting, this one.

This is not one of the controllers registered against `DevelopmentAdmin`, so it's not actually checked.

However, moving this to a `canInit()` at least makes it consistent and easy enough to move there if we need to

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [reply src/Dev/DevelopmentAdmin.php:191]: Oh! Love that lol. Yeah in that case it's good that we're now actually explicitly checking this - it looks like it should have been that way the whole time, given that test case.

File: src/ORM/DatabaseAdmin.php
Comment (GuySartorelli) [reply src/ORM/DatabaseAdmin.php]: Oohhhh good shout. Yeah it'd be good if when/if we register it everything's all in place and it "just works".

File: src/Dev/DevelopmentAdmin.php
Comment (GuySartorelli) [reply src/Dev/DevelopmentAdmin.php:198]: Fair point. Admin will still be able to see them so there's no scenario where a link _was_ visible before and _won't_ be anymore. I think you're right, this is the better outcome.

**Comment by GuySartorelli** [comment]
Looking good! Do you have any thoughts about "see" vs "view", and "run" vs "execute" as per https://github.com/silverstripe/silverstripe-framework/pull/10979#pullrequestreview-1698226313? I think it might be good to use consistent language there but It's not a deal breaker either way IMO.

**Comment by andrewandante** [comment]
> Do you have any thoughts about "see" vs "view", and "run" vs "execute

Yeah it should definitely be cleaned up. I think "view" and "execute" feel the most correct. Will consolidate and squash

**Comment by andrewandante** [comment]
image

With the adjustment to priority 80 in [the GraphQL module PR](https://github.com/silverstripe/silverstripe-graphql/pull/559)

**Comment by GuySartorelli** [comment]
I was just running this locally as a last double check before merging and I noticed a big problem in live mode.

As an admin user, when I go to any of the /dev/* endpoints, I still get the "Confirm potentially dangerous action" screen - which is good
Confirm potentially dangerous action

However, as a non-admin user who has been granted access via the new permissions, I _do not_ get that screen. That screen should always appear in live mode for users who have permission to access the endpoint.

**Comment by andrewandante** [comment]
Interesting - it looks like the way this works is that it only triggers the confirmation if the user has a permission that it thinks is behind the URL - if I'm understanding correctly, this means this list: https://github.com/silverstripe/silverstripe-framework/blob/5/_config/requestprocessors.yml#L100-L101

I could add the list of new permissions to this but it feels a bit flaky? It means that each time a new task/permission or whatever gets added, that would need to also be added to the config - and would be easy to forget. The alternative would be to add some sort of wildcard check, rather than "just" `ADMIN`

I'm not certain on the best approach to this. It feels like writing a `/dev` middleware is overkill but that _would_ allow me to overload `hasAccess()` with a `DevelopmentAdmin::canInit()` check instead? Do you have a feeling for the approach?

**Comment by GuySartorelli** [comment]
We wouldn't be able to set the `AffectedPermissions` in a way that's useful because it's a blanket middleware. If we added all the new permissions, then it would always ask for confirmation if someone only had access to `/dev/config` but was trying to access `/dev/build` for example.

Possible solutions:
1. Subclass `URLSpecialsMiddleware` or `PermissionAwareConfirmationMiddleware` (or even just add a new simpler middleware since the permission checks are more complex than those middleware classes are intended to handle) specifically for the `/dev/*` endpoints
1. Add a new property to `PermissionAwareConfirmationMiddleware` which gives it a callback for checking permissions. It can then go "if the member has the permission code, _or_ if the callback returns true, ask for confirmation"
1. Have a distinct middleware singleton defined for each `/dev/*` endpoint, with the relevant permissions set (easy to do but ends up with config spaghetti and easy to forget for new endpoints)
1. Add the permission codes to a private static on the controller, and add a way to tell `PermissionAwareConfirmationMiddleware` to check the private static of whatever controller will eventually handle the request, and check for _those_ permissions.
1. Uhhhh...?? Maybe move the confirmation thing out of middleware into the controllers, or into some service that can be called from the controllers? I don't like that idea though.

In either of 1 or 2 you're gonna need a way to call the relevant `canInit()` method from within the middleware - and it can't just be the `DevelopmentAdmin::canInit()`, it needs to be the `canInit()` for the relevant controller, or else you'll still get into the scenario I described above where someone with _one_ permission gets asked for confirmation for an endpoint they _don't_ have permission to access.

Of all of these I think I like 4 the most, as it feels like the least spaghetti solution. It also has the added bonus of making the permission codes for a given `/dev/*` endpoint configurable for mad scientists.

In some of these scenarios, if `canInit()` is overridden in a project to include some check that _isn't_ directly permission-based, they may get into a position where they again have users not being asked for confirmation. I think we can just call that out of scope though.

**Comment by andrewandante** [comment]
Alright, after a lot of experimentation, I've added a specific DevAdmin middleware. Determining the final controller was too difficult inside the existing middleware implementation so this was cleanest I think, and should "absorb" additional `/dev` endpoints as they are added.

**Comment by GuySartorelli** [comment]
Looking good - but there's still a few more things to tidy up

## Problem running some tasks

When I run a dev/task with someone who _does_ have permissions for `/dev/tasks`, I get the output from running the task, but I _also_ get a "you don't have access to this page" message:

image

That doesn't happen with all tasks - I only tried the "Deletes all temporary test databases" task (which gives that output) and the "Login Session Garbage Collection Task" (which works correctly and gives the normal output).

If might just be that the "Deletes all temporary test databases" task doesn't actually work for arbitrary users, and needs a `canView()` applied.

## `/dev/build` not protected

`/dev/build` is listed as an explicit exception to the new middleware in `dev_urls-confirmation-exceptions` config, and is instead protected by the `url_specials-middleware` config block. We probably want to move it _out_ of the URL specials middleware and into the new middleware.

Currently I get no confirmation screen for `/dev/build`

## sub-urls not protected

Without this PR, running a `BuildTask` (e.g. `/dev/task/CleanupTestDatabasesTask`) or going to `/dev/graphql/build` prompts a confirmation - but with the PR, there's no confirmation.

**Comment by andrewandante** [comment]
Thanks again for your thorough testing!

For the first point, the TestDatabases task has a specific permissions check for `ADMIN` inside the run method, which is what throws that. Refactoring to a canView() is probably the way to go for that specific scenario.

For the `dev/build`, I think moving it across will largely work, though given the special edge-cases around it I'll likely have to explicitly code those into the middleware. Might be best anyway!

Last point - good point, will check in on that. Probably need to find some way to check permissions on the parent. Phew!

**Comment by andrewandante** [comment]
Hmm. So I'm working on the dev/build endpoint, and for a reason I haven't tracked down yet, it wants to move me to `stage=Stage` - which fails inside `VersionedHTTPMiddleware::checkPermissions()`  (because of `Versioned::can_choose_site_stage()`). This means if the user doesn't _also_ have CMS permissions, they can't run the `dev/build`. I can fix it by adding `CAN_DEV_BUILD` to `Versioned::non_live_permissions` but that doesn't feel right? Maybe I'm overthinking it, but at that point the `$request->getURL()` is _not_ `/dev/build` so I can't use the same logic we are using in other places. Will keep digging but if there's something immediately obvious I'd love to hear it 😅 

Edit: it's coming from `VersionedStateExtension::updateLink()` - I guess the ReadingMode is `stage` when you hit a `/dev` URL?
Edit2: yep, inside `DevelopmentAdmin::init()`. In that case I think I _do_ need to add `CAN_DEV_BUILD` to the permission lists for Versioned.

**Comment by andrewandante** [comment]
Alright I think with all that, I've resolved those three concerns

**Comment by GuySartorelli** [comment]
Awesome, nice work! I'll take a proper look on Monday and put it through its paces.

File: src/Dev/Tasks/CleanupTestDatabasesTask.php
Comment (GuySartorelli) [inline src/Dev/Tasks/CleanupTestDatabasesTask.php:26]: We should probably keep this (just replace it with `if (!$this->canView()`) so we're still checking permissions for any weird scenario where this gets called directly, and giving the same result we were giving before.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
This is the only change needed, at a glance.

**Comment by GuySartorelli** [comment]
It's veeeeery nearly there. All of the permissions are working great, now the only problem is the one from https://github.com/silverstripe/silverstripe-framework/pull/10979#issuecomment-1788051188 where I still see "Running Task Deletes all temporary test databases".

The task doesn't _actually_ run because the `run()` method exits out early due to the condition in that method, but the runner itself it still calling that method which is a problem.
The task is also still visible in the `/dev/tasks` list, which it shouldn't be. I think? I keep getting confused about when canview is meant to work in this scenario - and I think that confusion is also a bad thing. Probably we should just make it so that the canview checks in tasks are always respected, because if I'm getting confused then other people will also get confused.

**Comment by andrewandante** [comment]
Ah yep, I see the issue there. The "can view all tasks" permission means that you can see the task, but the `canView()` check on the task itself says you can't.

My instinct here is to actually update the `canView()` check for this task to include the parent permission, though I don't know if that's the intention of the check for the task in the first place. Headache 🤔 

Actually it kind of feels like the `init()` on each `BuildTask` should be checking `canView()`, but that feels like it's going to introduce an API break...

**Comment by GuySartorelli** [comment]
> Actually it kind of feels like the `init()` on each `BuildTask` should be checking `canView()`, but that feels like it's going to introduce an API break...

There is no `BuildTask::init()`.... not sure what you mean?

I think the `getTasks()` and `runTask()` methods in `TaskRunner` should do this before returning/running a task:
```php
if ($task->hasMethod('canView') && !$task->canView()) {
    // skip, disallow, can't view it can't run it, not allowed.
}
```

That logic could probably be put into the `taskEnabled` method instead, which is already called from `getTasks()` and should _probably_ be called in `runTask()` instead of calling `$inst->isEnabled()` directly.

This should be fine in terms of backwards compatibility, because a `canView()` method on a `BuildTask` hasn't meant anything up until now. We're free to start calling it now and acting on its response.

**Comment by andrewandante** [comment]
> There is no `BuildTask::init()`.... not sure what you mean?

Exactly, an API break 😉 

Yeah I think with what you've proposed it means the task overrides the global permission, which is probably more intuitive. I might have to update one of the descriptions to reflect that, but it does make sense.

**Comment by GuySartorelli** [comment]
Oh, you were meaning you proposed introducing a new method? It's not an API break to add new methods so long as those methods aren't abstract. But in this case I don't think we need a new init method.

**Comment by andrewandante** [comment]
Test failures seem unrelated to my changes

File: src/Dev/TaskRunner.php
Comment (GuySartorelli) [inline src/Dev/TaskRunner.php:1]: This is a lot closer to ready now, except because `getTasks()` is skipping disabled and not-viewable tasks, I get the message `The build task "CleanupTestDatabasesTask" could not be found` when trying to run `/dev/tasks/CleanupTestDatabasesTask` (when I don't have permissions for it) - but I should see `The task is disabled or you do not have sufficient permission to run it`

It looks like this was already giving the wrong message for disabled messages, so we _could probably_ say getting the message right is out of scope. Overall the behaviour now is very good, and any changes to messaging etc won't kick in until people start applying `canView()` to tasks and giving people these new permissions, so overall I think it's a good change as-is.

I'll let you determine if you want to try get this messaging 100% correct, but I'm happy to merge as-is if you want that to be treated separately given it was (at least partially) a pre-existing issue.

Comment (andrewandante) [reply src/Dev/TaskRunner.php:1]: Oh that's a good point - I think tidying up the messaging makes sense, while I'm here, so I'll do that.

**Review by GuySartorelli — APPROVED** [review]
Great, thanks!

MariaDB failures are unrelated - see https://github.com/silverstripe/silverstripe-graphql/issues/561 and https://github.com/silverstripe/silverstripe-framework/issues/11046

**Pull request merged**
